### DISCUSSION TITLE
how to set webp's loop count if the original picture is a gif with GIF87a type?

### DISCUSSION THREAD

**Discussion by t00350320** [libvips/libvips#3632] [General]
hi ,
We have a task to transfer a gif(GIF87a, gif1.zip) [gif1.zip](https://github.com/libvips/libvips/files/12451174/gif1.zip) to webp, 
here is our govips demo code:
```
func main() {
        vips.Startup(nil)
        defer vips.Shutdown()
        p:=vips.NewImportParams()
        p.NumPages.Set(-1)
        imagePath := "gif1.gif"
        buf, err := ioutil.ReadFile(imagePath)  
        imageref,err := vips.LoadImageFromBuffer(buf,p)
        if err !=nil { 
          fmt.Println(err)
        }
        checkError(err)
        //imageref.RemoveMetadata()


        ep := vips.NewDefaultWEBPExportParams()
        image1bytes, _, err := imageref.Export(ep)
        err = ioutil.WriteFile("output.webp", image1bytes, 0644)
        checkError(err)

}
```
in order to get a thin image, we default use "RemoveMetadata()" before Export to webp， the output webp's loop count is 0 , so the webp loop forever.
```
./webpinfo -bitstream_info webp1.webp
Chunk ANIM at offset     30, length     14
  Background color:(ARGB) ff ff ff ff
  Loop count      : 0
```
if we remove  "RemoveMetadata()" , the output is 1, the webp loop once
```
Chunk ANIM at offset     30, length     14
  Background color:(ARGB) ff ff ff ff
  Loop count      : 1
```
So,webp "Loop count:" parameter differs with the one that execute "RemoveMetadata()"
,  two doubts here:
a:) why RemoveMetadata() affects "Loop count : 0" parameter? 
b:) how to set  webp's loop count if the original picture is a gif with GIF87a type? In other words, if there is no loop count for GIF87a, how to confirm the webp's loop count?

We also tested the normal GIF89a type(gif2.zip) picture with loop 2, it works well when transfered to webp.
[gif2.zip](https://github.com/libvips/libvips/files/12451175/gif2.zip)

**Environment**
(please complete the following information)
 - OS: [Linux version 3.10.0-1062.el7.x86_64  (gcc version 4.8.5 20150623]

 - Vips: [e.g. 8.14.2]

**Additional context**
gif may has two types: GIF87a, GIF89a. While 87a has no reference to loop count and 89a's loop count is set near "0x31d". 
For example, 89a type gif (gif2.zip), “31D: 02 00 # Loop 2 times".

gif89a doc，http://www.vurdalakov.net/misc/gif/netscape-looping-application-extension。

thank you very much!!!

**Comment by jcupitt** [comment]
Hi @t00350320,

I'd say this was a bug in govips -- `RemoveMetadata` should not remove `loop`.

There are a number of other header fields it should probably not remove, for example:

```
$ vipsheader -a 3198.gif 
3198.gif: 298x193 uchar, 3 bands, srgb, gifload
width: 298
height: 193
bands: 3
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 1
yres: 1
filename: 3198.gif
vips-loader: gifload
n-pages: 139
loop: 0
delay: 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30 30
background: 104 54 42 
gif-palette: -14010776 -4605511 -12035707 -16119469 -14282072 -13424976 -3486773 -15529865 -15200121 -15331208 -15267705 -15657429 -13088141 -10194021 -13555055 -15726488 -9934744 -13483158 -12373841 -8943728 -14339754 -6642007 -9140821 -13362487 -2631721 -1461003
bits-per-sample: 8
```

So it should leave at least `n-pages`, `loop`, `delay`, `background`, `gif-palette`, and `bits-per-sample`. There are others used by other savers to set various parameters, `vips/header.h` has a list.

Another option would be for you to pass the `strip` option to the saver. This will prevent extra metadata being written to the file during save.

**Reply by t00350320** [reply]
pretty thanks for your proposal. will try the "strip" option . 
But with GIF89a type(gif2.zip) picture (with loop 2), it works well when transfered to webp even with RemoveMetadata(). this makes me very confused
